[PrintAsObjC] Disable -Watimport-in-framework-header warning in generated headers

beccadax opened the pull request:
This is a new warning in clang when you use @import in a framework header. That’s often a mistake, but it isn’t in our generated `-Swift.h` headers. If we’re building with a clang that supports this warning, we’ll now emit a pragma to disable it.

Fixes rdar://41552597.

beccadax committed:
[ClangImporter] Run arc-conventions test with built clang

Important when we’re testing features which are only in the new clang.

beccadax committed:
Add -Watimport-in-framework-header flag to .h files

Silences a new warning in clang.

jrose-apple commented on test/PrintAsObjC/imports.swift:
```diff
@@ -5,10 +5,16 @@
 // RUN: %FileCheck %s < %t/imports.h
 // RUN: %FileCheck -check-prefix=NEGATIVE %s < %t/imports.h
-// RUN: %check-in-clang %t/imports.h -I %S/Inputs/custom-modules/ -F %S/Inputs/
+// RUN: %check-in-clang %t/imports.h -I %S/Inputs/custom-modules/ -F %S/Inputs/ -Watimport-in-framework-header
```
Did you verify that the warning fires without your fix here? I think it really does have to look like a framework header for it to show up. We should make sure we're testing the right thing.

brentdax replied:
You're right, clang looks up path it's compiling for something with a .framework extension. I'll check if one of the other tests is more suitable.

jrose-apple replied:
It's fine to add a new RUN line, too; sometimes that's lighter weight than adding a whole new test.

beccadax committed:
Test new warning in a framework-y environment

The atimport-in-framework-header warning is only actually emitted if the header’s path looks like it’s in a framework. We don’t have a test like that, but mixed-framework.swift probably *ought* to be like that, so if we make it more realistic, it will suit our purposes.

jrose-apple commented on test/PrintAsObjC/mixed-framework.swift:
```diff
@@ -1,17 +1,20 @@
-// RUN: %FileCheck -check-prefix=CHECK -check-prefix=HEADER %s < %t/mixed-header.h
-// RUN: %check-in-clang -include %S/Inputs/Mixed.framework/Headers/Mixed.h %t/mixed-header.h
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -module-name Mixed -import-objc-header %S/Inputs/Mixed.framework/Headers/Mixed.h %s -typecheck -emit-objc-header-path %t.framework/mixed-header.h
+// RUN: %FileCheck -check-prefix=CHECK -check-prefix=HEADER %s < %t.framework/mixed-header.h
```
Nitpick: Did you want to put this in the Headers directory?

brentdax replied:
I don't think so, but I could be wrong. My understanding is that the first header it generates is the one that might actually be packaged in a framework, while the second is one that might be generated as a temporary file when using the framework. But I could be misunderstanding the test; I'm not too familiar with why you would or wouldn't pass `-import-underlying-module -parse-as-library` when doing `-emit-objc-header-path`.

jrose-apple replied:
There's only one generated header; the first header listed is effectively a bridging header. (If you wanted to closely more emulate a mixed-source framework, you could provide a module map and use `-import-underlying-module` instead. But then only `public` things get printed.)

Oh, and the generated header is always considered part of a framework build product.

brentdax replied:
Okay, so we're not really simulating anything very specifically here. Might as well move the header, then.

beccadax committed:
Move test header into a more tidy spot

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@jrose-apple Okay with merging this?

jrose-apple commented:
Ah, yeah, sorry, those were my only remaining comments.

jrose-apple commented:
Thanks for taking this!

brentdax commented:
@swift-ci please smoke test and merge

brentdax commented:
@swift-ci please smoke test linux platform

shahmishal commented:
@swift-ci please smoke test linux platform
